[java] New rule: JUnit5TestShouldBePackagePrivate

ajeans opened the pull request:
Add the rule to check that JUnit5 tests methods and classes are package private.
Add unit tests

## Describe the PR

Propose a codestyle rule for Jupiter / JUnit5 tests according to the feedback received from @adangel in the issue.

Not sure how to add the 'externalInfoURL', so simply pointed at the rule name, assuming that the site is generated from the XML rule file

## Related issues

- Fixes #3239 
- Relates #3288 - that's another rule to handle private tests

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [X] Added unit tests for fixed bug/feature
- [X] Passing all unit tests
- [X] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [X] Added (in-code) documentation (if needed)

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,76 @@ public class MyTest {
+        </properties>
+        <example>
+            <![CDATA[
+class MyTest {
```
Please add comments in the example so that we know what to look at. Eg
```suggestion
class MyTest { // not public, that's fine
```

You could also reduce the size of the example by removing the body of methods and making them single-line. Eg
```java
    @Test
    public void testBad() {  } // should not have a public modifier
```
and so on

ajeans replied:
Done

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,76 @@ public class MyTest {
+]]>
+                </value>
+            </property>
+            <property name="testClassPattern" type="Regex" description="The regex pattern used to identify test classes" value="Test" />
```
We're looking for a junit 5 test class, ie a class in which some method has one of the test annotations mentioned above. I don't think there's a need to check for the name of the class?

ajeans replied:
Well I followed what was in `JUnit4TestShouldUseTestAnnotation` but I agree I don't see the rationale for it.

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This changeset changes 0 viola...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset changes 0 violations,<br />
introduces 2268 new violations, 15 new errors and 0 new configuration errors,<br />
removes 1877 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/691e5d2eda6d42b0b54ce18de5abe831/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,76 @@ public class MyTest {
+    <rule name="JUnit5TestShouldBePackagePrivate"
+          language="java"
+          since="6.35"
+          message="JUnit 5 tests that use @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest should be package private."
```
```suggestion
          message="JUnit 5 tests should be package-private."
```

I don't know of Junit 5 tests that don't use one of these annotations.

ajeans replied:
Done

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,76 @@ public class MyTest {
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#junit5testshouldbepackageprivate">
+        <description>
+            JUnit 5 tests should be package private at the class level and for each method that uses @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest.
```
Please break this up into several lines. Also this should include a line to explain why the rule exist. Why should junit 5 tests be package private? Why not Junit 4 tests?

ajeans replied:
Done

oowekyala commented on pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnit5TestShouldBePackagePrivate.xml:
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I think this test file should also ensure that we don't flag junit 4 tests.

ajeans replied:
Done

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,76 @@ public class MyTest {
 
+    <rule name="JUnit5TestShouldBePackagePrivate"
+          language="java"
+          since="6.35"
```
```suggestion
          since="6.35.0"
```

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,76 @@ public class MyTest {
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//ClassOrInterfaceDeclaration[
```
In XPath 2 you can merge both expressions, like

```ruby
//ClassOrInterfaceDeclaration[...]/(
  self::*[@Abstract=false() and @PackagePrivate=false()]
| ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/...
)
```

Please do this, as much of the rule is duplicated and it's hard to see what's going on. 

Also please report the violation on the MethodDeclaration and not the ClassOrInterfaceBodyDeclaration. Otherwise the line number of the violation is on the annotation...

ajeans replied:
While I managed to report the violation on MethodDeclaration, I don't see how to reduce the two paths in one:
* the first one reports one violation on the class (if not package-private nor abstract) for any number of `@Test` methods contained in the class
* the second one reports one violation per `@Test` methods (if not package-private)

Could you elaborate? I'm probably missing something...

ajeans replied:
While I managed to report the violation on MethodDeclaration, I don't see how to reduce the two paths in one:

    the first one reports one violation on the class (if not package-private nor abstract) for any number of @Test methods contained in the class
    the second one reports one violation per @Test methods (if not package-private)

Could you elaborate? I'm probably missing something...

oowekyala replied:
This is roughly what I meant:
```ruby
//ClassOrInterfaceDeclaration[
      (: a Junit 5 test class, ie, it has methods with the annotation :)
      ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/Annotation[pmd-java:typeIs(...)]
]/(
    self::*[@Abstract=false() and @PackagePrivate=false()] 
|   ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration[Annotation[pmd-java:typeIs(...)]]
                        /MethodDeclaration[@PackagePrivate=false()]
)
```

So there's still two paths, but with a common ancestor.

ajeans replied:
Thanks, I think I got it, please look at the updated rule.

oowekyala reviewed (commented):
Welcome and thanks for your work on this! I have a couple of suggestions for improvements, please take a look and push new commits when you can

oowekyala commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,67 @@ public class MyTest {
+        <description>
+JUnit 5 tests should be package private at the class level and for each method that uses @Test, @RepeatedTest,
+@TestFactory, @TestTemplate or @ParameterizedTest.
+Contrary to JUnit4 tests that required public visibility to be run by the engine, JUnit5 tests may be run
```
Thanks for updating the description! Minor suggestion:

> JUnit5 tests may be run only with package-private visibility

This sounds like they *require* package-private visibility. IMO a better way to put it would be "JUnit5 tests can also be run if they're package-private"

ajeans replied:
Done

adangel commented on pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/JUnit5TestShouldBePackagePrivate.xml:
```diff
@@ -0,0 +1,101 @@
+
+    <test-code>
+        <description>Public modifier is not necessary on a test method nor on the class</description>
+        <expected-problems>2</expected-problems>
```
with multiple problem, we can pin it down to a specific line-number to ensure, we don't flag something wrong.
here it would be

```xml
<expected-linenumbers>3,5</expected-linenumbers>
```

The expected-linenumbers needs to be added additionally to expected-problems.

adangel commented on pmd-java/src/main/resources/category/java/bestpractices.xml:
```diff
@@ -802,6 +802,68 @@ public class MyTest {
+          message="JUnit 5 tests should be package-private."
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#junit5testshouldbepackageprivate">
```
> Not sure how to add the 'externalInfoURL', so simply pointed at the rule name, assuming that the site is generated from the XML rule file

That looks fine. The rule name is all lowercase. We also have a check which fails the build if the url is not correct.

adangel commented:
@ajeans This new rule detects two violations in the spring-framework: https://chunk.io/pmd/0094edd7ef3f417898be43865c65ec41/diff/spring-framework/index.html#section-violations
* https://github.com/spring-projects/spring-framework/tree/v5.0.6.RELEASE/spring-test/src/test/java/org/springframework/test/context/junit/jupiter/FailingBeforeAndAfterMethodsSpringExtensionTestCase.java#L246
* https://github.com/spring-projects/spring-framework/tree/v5.0.6.RELEASE/spring-test/src/test/java/org/springframework/test/context/junit/jupiter/FailingBeforeAndAfterMethodsSpringExtensionTestCase.java#L260

Both are private static inner classes. Are these valid cases?

Since they are private, I thought, they should only be picked up by the other (to be implemented) rule JUnit5TestNoPrivateModifier...
We should avoid to report these cases twice.

Besides that, these two cases might be a false positive for JUnit5TestNoPrivateModifier anyway - but probably not so easy to detect. It seems, that they use `org.junit.platform.engine.discovery.DiscoverySelectors`, which seems to allow private test classes...

So for this rule, I'd suggest to exclude private classes/methods, as this rule should only flag cases, where the class/method is public/protected. This should be possible by restricting the query additionally via `[@Private=false()]`.

ajeans commented:
Thanks for the feedbacl @adangel 

> @ajeans This new rule detects two violations in the spring-framework: https://chunk.io/pmd/0094edd7ef3f417898be43865c65ec41/diff/spring-framework/index.html#section-violations
> 
>     * https://github.com/spring-projects/spring-framework/tree/v5.0.6.RELEASE/spring-test/src/test/java/org/springframework/test/context/junit/jupiter/FailingBeforeAndAfterMethodsSpringExtensionTestCase.java#L246
> 
>     * https://github.com/spring-projects/spring-framework/tree/v5.0.6.RELEASE/spring-test/src/test/java/org/springframework/test/context/junit/jupiter/FailingBeforeAndAfterMethodsSpringExtensionTestCase.java#L260
> 
> 
> Both are private static inner classes. Are these valid cases?

It appears they are, those are inner static classes annotated with `@Test` and fed through a `@TestFactory` to test a JUnit extension. 

> 
> Since they are private, I thought, they should only be picked up by the other (to be implemented) rule JUnit5TestNoPrivateModifier...
> We should avoid to report these cases twice.

Agreed, I added two unit tests and changed the rule to target protected and public modifiers, and ignore private modifiers.

> Besides that, these two cases might be a false positive for JUnit5TestNoPrivateModifier anyway - but probably not so easy to detect. It seems, that they use `org.junit.platform.engine.discovery.DiscoverySelectors`, which seems to allow private test classes...

That's using the jupiter engine dependency directly, so definitely advanced usage. The rule is targeted at devs that simply use the test annotations for simple and documented use cases. If you are doing the test discovery using an hardcoded list of test classes, you're way outside the junit user guide, and this rule doesn't make a lot of sense...
We can have this conversation when we implement the second rule in a separate MR but I would argue that this is an advanced usage and people should simply annotate the class to ignore the rule.

> So for this rule, I'd suggest to exclude private classes/methods, as this rule should only flag cases, where the class/method is public/protected. This should be possible by restricting the query additionally via `[@Private=false()]`.

Done with latest commit, hope this is what you had in mind.

adangel reviewed (approved):
Thanks! Looks good to me!
